docs/stores/settings.md
# Vext Settings Store

> Pinia store that holds settings for different VEXT tools (brush & shape).

## State

| Name      | Description                      | Type   | Values | Default              |
| --------- | -------------------------------- | ------ | ------ | -------------------- |
| pointerMenu | whether the pointer menu is open | boolean | -      | `false`         |
| onAction | whether to open the pointer menu after an annotation action | boolean | -      | `true`         |
| onGesture | whether to open the pointer menu after with a specific gesture | boolean | -      | `true`         |
| closeOnClick | whether to close the pointer menu after clicking one of its items | boolean | -      | `true`         |
| historyLimit | how many interactions to store in the history log | number | -      | `50`         |
| activeColor | currently active color (primary or secondary) | number | `[0,1]`      | `0`         |
| color0 | primary color | string | -      | `"#ffffff"`         |
| color1 | secondary color | string | -      | `"#000000"`         |
| brushSize | size of the drawing brush | number | -      | `1`         |
| brushDecimation | how much to decimate drawn lines | number | -      | `5`         |
| shape | selected shape in the shape mode | string | `["text", "circle", "rectangle", "triangle"]`      | `"text"`         |
| shapeDim0 | size of the first dimension for the new shape | number | -      | `10`         |
| shapeDim1 | size of the second dimension for the new shape | number | -      | `10`         |
| defaultColors | array of colors to use for layers | number | -      | `["#4e79a7","#f28e2c","#e15759","#76b7b2","#59a14f","#edc949","#af7aa1","#ff9da7","#9c755f","#bab0ab"]`         |

## Getters

### color

> Returns the currently selected color string.

## Methods

### emit

> Emit an event from the note event handler.
> <br>`@param {String}` name - event name
> <br>`@param {*}` data - event payload

### on

> Register an event handler.
> <br>`@param {String}` name - event name
> <br>`@param {Function}` hadler - event callback handler
> <br>`@returns` handle id

### off

> Remove an event handler for the given event.
> <br>`@param {String}` name - event name
> <br>`@param {Number}` handler - event handle
> <br>`@returns` true if the event was removed

### defaultColorAt

> Returns the default color at a specific index.
> Colors are wrapped -> starts at the beginning in index is out of bounds.
> <br>`@param {Number}` index - color index
> <br>`@returns {String}` the color

### setHistoryLimit

> Sets the history interaction limit (minimum value of 1 is enforced)
> <br>`@param {Number}` limit

### setBrushSize

> Sets the brush size to use for drawing.
> <br>`@param {Number}` size - brush size in pixels
> <br>`@param {Boolean}` record - whether to record this action (default `true`)

### setBrushDecimation

> Sets the brush decimation value.
> <br>`@param {Number}` value - value in pixels
> <br>`@param {Boolean}` record - whether to record this action (default `true`)

### selectColor

> Select a color by id (either primary = 0 or secondary = 1).
> If the passed id is not 0 or 1, the selected color is toggled.
> <br>`@param {Number}` id - color id
> <br>`@param {Boolean}` record - whether to record this action (default `true`)

### setColor

> Sets the color to use for the currently active color.
> <br>`@param {String}` color - color value
> <br>`@param {Boolean}` record - whether to record this action (default `true`)

### setColorPrimary

> Sets the color to use for the primary color.
> <br>`@param {String}` color - color value
> <br>`@param {Boolean}` record - whether to record this action (default `true`)

### setColorSecondary

> Sets the color to use for the secondary color.
> <br>`@param {String}` color - color value
> <br>`@param {Boolean}` record - whether to record this action (default `true`)

### setShape

> Sets the shape to use in the shape mode
> <br>`@param {String}` shape - shape type
> <br>`@param {Boolean}` record - whether to record this action (default `true`)


docs/stores/note-settings.md
# Vext Note Settings Store

> Pinia store that holds settings for different VEXT tools (brush & shape).

## State

| Name      | Description                      | Type   | Values | Default              |
| --------- | -------------------------------- | ------ | ------ | -------------------- |
| key | key down information | object | -      | `{}`         |

## Methods

### alert

> Set the alert text and type for a global alert
> <br>`@param {String}` text
> <br>`@param {String}` type



src/store/settings.js
import { defineStore } from "pinia"
import { useVextHistory } from "./history";
import EventHandler from "@/use/event-handler";
import { ACTIONS } from "@/use/enums";

const EVENTS = new EventHandler();

const vextSettingsStore = {

    state: () => {
        return {

            pointerMenu: false,
            onAction: true,
            onGesture: true,
            closeOnClick: true,

            historyLimit: 50,

            activeColor: 0,
            color0: "#ffffff",
            color1: "#000000",
            brushSize: 1,
            brushDecimation: 5,

            shape: "text",
            shapeDim0: 10,
            shapeDim1: 10,

            // d3.schemeTableau10
            defaultColors: [
                "#4e79a7",
                "#f28e2c",
                "#e15759",
                "#76b7b2",
                "#59a14f",
                "#edc949",
                "#af7aa1",
                "#ff9da7",
                "#9c755f",
                "#bab0ab"
            ],
        };
    },

    getters: {
        color: state => state.activeColor === 0 ? state.color0 : state.color1,
    },

    actions: {

        emit(name, data) {
            EVENTS.emit(name, data);
        },

        on(name, handler) {
            return EVENTS.on(name, handler);
        },

        off(name, handler) {
            return EVENTS.off(name, handler);
        },

        defaultColorAt(index) {
            return this.defaultColors[index % this.defaultColors.length]
        },

        setHistoryLimit(limit) {
            const history = useVextHistory();
            history.limit = Math.max(1, Math.floor(limit));
        },

        setBrushSize(size, record=true) {
            const newVal = Math.max(size, 1);
            if (record) {
                const history = useVextHistory();
                history.do("set brush size to "+newVal,
                    this.setBrushSize.bind(this, newVal, false),
                    this.setBrushSize.bind(this, this.brushSize, false)
                );
            }
            this.brushSize = newVal;
            this.emit(ACTIONS.BRUSH_SIZE, newVal);
        },

        setBrushDecimation(value, record=true) {
            const newVal = Math.max(value, 0);
            if (record) {
                const history = useVextHistory();
                history.do("set brush decimation to "+newVal,
                    this.setBrushDecimation.bind(this, newVal, false),
                    this.setBrushDecimation.bind(this, this.brushDecimation, false)
                );
            }
            this.brushDecimation = newVal;
            this.emit(ACTIONS.BRUSH_DECIMATE, newVal);
        },

        selectColor(id=null, record=true) {
            const newColor = id === 0 || id === 1 ? id : Math.abs(1-this.activeColor);
            if (record) {
                const history = useVextHistory();
                history.do("select color to "+(newColor === 0 ? 'primary' : 'secondary'),
                    this.selectColor.bind(this, newColor, false),
                    this.selectColor.bind(this, this.activeColor, false)
                );
            }
            this.activeColor = newColor
            EVENTS.emit(ACTIONS.COLOR_CHANGE, newColor);
        },

        setColor(color, record=true) {
            if (this.activeColor === 0) {
                this.setColorPrimary(color, record);
            } else {
                this.setColorSecondary(color, record);
            }
        },

        setColorPrimary(color, record=true) {
            if (record) {
                const history = useVextHistory();
                history.do("set primary color to "+color,
                    this.setColorPrimary.bind(this, color, false),
                    this.setColorPrimary.bind(this, this.color0, false),
                );
            }
            this.color0 = color;
            if (this.activeColor === 0) {
                EVENTS.emit(ACTIONS.COLOR_VALUE, color);
            }
        },

        setColorSecondary(color, record=true) {
            if (record) {
                const history = useVextHistory();
                history.do("set secondary color to "+color,
                    this.setColorSecondary.bind(this, color, false),
                    this.setColorSecondary.bind(this, this.color1, false),
                );
            }
            this.color1 = color;
            if (this.activeColor === 1) {
                EVENTS.emit(ACTIONS.COLOR_VALUE, color);
            }
        },

        setShape(shape, record=true) {
            switch (shape) {
                case "text":
                case "circle":
                case "rectangle":
                case "triangle":
                    if (shape !== this.shape) {
                        this.shape = shape;
                        if (record) {
                            const history = useVextHistory();
                                history.do("set shape to "+shape,
                                this.setShape.bind(this, shape, false),
                                this.setShape.bind(this, this.shape, false),
                            );
                        }
                    }
                    break;
            }
        }

    },
}

const useVextSettings = defineStore("vext-settings", vextSettingsStore);

export { useVextSettings, vextSettingsStore }